libfreefare/freefare.h

```diff
@@ -521,6 +521,28 @@ uint8_t		*tlv_decode(const uint8_t *istream, uint8_t *type, uint16_t *size);
 size_t		tlv_record_length(const uint8_t *istream, size_t *field_length_size, size_t *field_value_size);
 uint8_t		*tlv_append(uint8_t *a, uint8_t *b);
 
+typedef enum mifare_key_type {
+    MIFARE_KEY_DES,
+    MIFARE_KEY_2K3DES,
+    MIFARE_KEY_3K3DES,
+    MIFARE_KEY_AES128,
+
+    MIFARE_KEY_LAST = MIFARE_KEY_AES128
```

Comment:
What is this `MIFARE_KEY_LAST` used for?

Reply:
I wanted to make sure that if there was ever the need for it people wouldn't be tempted to assume MIFARE_KEY_AES128 was the numerically largest enumerated key type.

Reply:
I can't see an use-case for such value :thinking: and I am wondering if this would not become painful e.g. in a `switch` statement when the compiler wants all defined values to have a `case`?

Are you thinking about something like `ELAST` in `errno.h` (where the different values are `#defined`) for which I can guess use-cases (use `ELAST + 1`, `ELAST + 2` and so on in an application)?

Reply:
> I am wondering if this would not become painful e.g. in a switch statement when the compiler wants all defined values to have a case?

By making `MIFARE_KEY_LAST` equal to the numeric value of the last entry, we avoid problems like what you describe—since all of the numeric values would indeed be in use.

> Are you thinking about something like `ELAST` in errno.h (where the different values are #defined) for which I can guess use-cases (use `ELAST + 1`, `ELAST + 2` and so on in an application)?

Not really, I'm thinking more about the possibility of input sanity checks, among other things. LibNFC suffers from this problem with the `nfc_modulation_type` enum, where `NMT_DEP` has a dire warning in the header that it must always be the last item:

https://github.com/nfc-tools/libnfc/blob/master/include/nfc/nfc-types.h#L301-L315

Stuff like that sucks because then you have to renumber when adding new items, which breaks APIs, etc...

Was hoping to avoid the possibility of that ever happening by including the `*_LAST` enum from the start.

Reply:
Okay, I think I understand, thanks!  It may make sense if it is actually used then :smile: